File: backend/app/services/constituency_service.py
from typing import List, Optional
from app.repositories.constituency_repository import ConstituencyRepository
from app.repositories.person_repository import PersonRepository
from app.models.constituency import ConstituencyResponse, ConstituencyWinnerHistoryResponse, WinnerHistoryRecord

class ConstituencyService:
    def __init__(self, repository: ConstituencyRepository = None, person_repo: PersonRepository = None):
        self.repository = repository or ConstituencyRepository()
        self.person_repo = person_repo or PersonRepository()

    def list_constituencies(self, district_id: Optional[str] = None) -> List[ConstituencyResponse]:
        """
        Fetches all constituencies, optionally filtered by district_id.
        Maps raw data to ConstituencyResponse models.
        """
        if district_id:
            raw_constituencies = self.repository.get_constituencies_by_district(district_id)
        else:
            raw_constituencies = self.repository.get_all_constituencies()

        result = []
        for c in raw_constituencies:
            result.append(ConstituencyResponse(
                id=c.get("PK", ""),
                name=c.get("name", ""),
                district_id=c.get("district_id", ""),
                type=c.get("type", "")
            ))
            
        return result

    def _slugify(self, name: str) -> str:
        if not name:
            return ""
        import re
        s = name.lower()
        s = re.sub(r'[^a-z0-9\s]', '', s)
        s = s.strip()
        s = re.sub(r'\s+', '-', s)
        return s

    def _parse_margin(self, raw_margin: str) -> int:
        try:
            return int(str(raw_margin).replace(",", ""))
        except (ValueError, TypeError):
            return 0

    def _get_party_info(self, raw_party_id: str, party_cache: dict) -> dict:
        if not raw_party_id:
            return {"id": "NA"}

        if raw_party_id in party_cache:
            return party_cache[raw_party_id]

        party_data = self.repository.get_party_by_id(raw_party_id) or {}
        
        info = {
            "id": raw_party_id,
            "name": party_data.get("name"),
            "short_name": party_data.get("short_name"),
            "logo_url": party_data.get("logo_url"),
            "color_bg": party_data.get("color_bg"),
            "color_text": party_data.get("color_text"),
            "color_border": party_data.get("color_border")
        }
        party_cache[raw_party_id] = info
        
        return info

    def get_winner_history(self, constituency_id: str) -> ConstituencyWinnerHistoryResponse:
        """
        Fetches winner history for a specific constituency.
        """
        raw_history = self.repository.get_winner_history(constituency_id)
        
        # Batch fetch person metadata for consistent names/slugs
        person_ids = list(set(h.get("person_id") for h in raw_history if h.get("person_id")))
        persons = self.person_repo.get_persons_by_ids(person_ids)
        person_map = {p.get("PK"): p for p in persons if p.get("PK")}

        history_records = []
        party_cache = {}

        for h in raw_history:
            margin = self._parse_margin(h.get("winning_margin", 0))
            party_info = self._get_party_info(h.get("party_id"), party_cache)

            person_id = h.get("person_id")
            person_meta = person_map.get(person_id, {}) if person_id else {}
            
            # Favor name from Person record
            winner_name = person_meta.get("name") or h.get("candidate_name") or "Unknown"
            year_val = h.get("year") or 0
            if isinstance(year_val, str):
                year_val = int(year_val)
            
            history_records.append(WinnerHistoryRecord(
                year = year_val,
                winner=str(winner_name),
                profile_pic=person_meta.get("image_url") or h.get("profile_pic"),
                party=party_info,
                margin=margin,
                person_id=person_id,
                slug=self._slugify(winner_name) if winner_name != "Unknown" else None
            ))
            
        # Sort history by year descending to ensure latest winner is at index 0
        history_records.sort(key=lambda x: x.year, reverse=True)
            
        # Fetch constituency metadata for statistics
        metadata = self.repository.get_constituency_metadata(constituency_id)
        stats = []
        if metadata and "statistics" in metadata:
            raw_stats = metadata["statistics"]
            for year, data in raw_stats.items():
                stats.append({
                    "year": int(year),
                    "total_electors": data.get("total_electors", 0),
                    "total_votes_polled": data.get("total_votes_polled", 0),
                    "poll_percentage": data.get("poll_percentage", 0),
                    "male": data.get("male"),
                    "female": data.get("female"),
                    "third_gender": data.get("third_gender")
                })
        
        # Sort stats by year descending
        stats.sort(key=lambda x: x["year"], reverse=True)

        # Extract plain constituency name from the ID (e.g. CONSTITUENCY#kolathur -> kolathur)
        constituency_name = constituency_id.replace("CONSTITUENCY#", "")
        
        return ConstituencyWinnerHistoryResponse(
            constituency=constituency_name,
            history=history_records,
            stats=stats
        )

File: frontend/lib/services/constituency.service.ts
import { ConstituencyRepository } from "../repositories/constituency.repository";
import { PersonRepository } from "../repositories/person.repository";
import { CandidateRepository } from "../repositories/candidate.repository";
import { 
  ConstituencyResponse, 
  ConstituencyWinnerHistoryResponse, 
  WinnerHistoryRecord 
} from "@/types/models";
import { getPartyLogo } from "../utils/party-utils";
import { normalizeTotalAssets, normalizeCriminalCases } from "../utils/profile-normalizers";

export class ConstituencyService {
  private repository: ConstituencyRepository;
  private personRepo: PersonRepository;
  private candidateRepo: CandidateRepository;

  constructor(repository?: ConstituencyRepository, personRepo?: PersonRepository, candidateRepo?: CandidateRepository) {
    this.repository = repository || new ConstituencyRepository();
    this.personRepo = personRepo || new PersonRepository();
    this.candidateRepo = candidateRepo || new CandidateRepository();
  }

  async listConstituencies(districtId?: string): Promise<ConstituencyResponse[]> {
    let rawConstituencies;
    if (districtId) {
      rawConstituencies = await this.repository.getConstituenciesByDistrict(districtId);
    } else {
      rawConstituencies = await this.repository.getAllConstituencies();
    }

    const constituencies = rawConstituencies.map((c: any) => ({
      id: c.PK,
      name: c.name,
      slug: c.normalized_name || c.PK.replace("CONSTITUENCY#", ""),
      district_id: c.district_id,
      type: c.type,
    }));
    constituencies.sort((a: any, b: any) => (a.name || "").localeCompare(b.name || ""));
    return constituencies;
  }

  private slugify(name: string): string {
    if (!name) return "";
    return name
      .toLowerCase()
      .replace(/[^a-z0-9\s]/g, "")
      .trim()
      .replace(/\s+/g, "-");
  }

  private parseMargin(rawMargin: any): number {
    if (!rawMargin) return 0;
    try {
      return parseInt(String(rawMargin).replace(/,/g, ""));
    } catch (error) {
      return 0;
    }
  }

  private async getPartyInfo(partyId: string, partyCache: Record<string, any>) {
    if (!partyId) return { id: "NA" };
    if (partyCache[partyId]) return partyCache[partyId];

    const partyData = (await this.repository.getPartyById(partyId)) || {};
    const info = {
      id: partyId,
      name: partyData.name,
      short_name: partyData.short_name,
      logo_url: getPartyLogo(partyData.short_name) || partyData.logo_url,
      color_bg: partyData.color_bg,
      color_text: partyData.color_text,
      color_border: partyData.color_border,
    };
    partyCache[partyId] = info;
    return info;
  }

  async getWinnerHistory(constituencyId: string): Promise<ConstituencyWinnerHistoryResponse> {
    const rawHistory = await this.repository.getWinnerHistory(constituencyId);
    const personIds = Array.from(new Set(rawHistory.map((h: any) => h.person_id).filter((id: string) => id)));
    const persons = await this.personRepo.getPersonsByIds(personIds as string[]);
    const personMap = persons.reduce((acc: any, p: any) => {
      acc[p.PK] = p;
      return acc;
    }, {});

    const partyCache: Record<string, any> = {};
    const historyRecords: WinnerHistoryRecord[] = [];

    // Sort raw history by year descending manually before loop to identify latest winner easily
    const sortedRawHistory = [...rawHistory].sort((a, b) => (parseInt(b.year) || 0) - (parseInt(a.year) || 0));

    for (let i = 0; i < sortedRawHistory.length; i++) {
      const h = sortedRawHistory[i];
      const margin = this.parseMargin(h.winning_margin);
      const partyInfo = await this.getPartyInfo(h.party_id, partyCache);
      const personMeta = h.person_id ? personMap[h.person_id] || {} : {};
      const winnerName = personMeta.name || h.candidate_name || "Unknown";
      const year = parseInt(h.year) || 0;

      let winRate = personMeta.win_rate;
      let totalWins = personMeta.total_wins;
      let totalContested = personMeta.total_contested;

      // Only calculate win rate for the latest record (current MLA) if missing in metadata
      if (i === 0 && h.person_id && !winRate) {
        try {
          const personHistory = await this.candidateRepo.getPersonHistory(h.person_id);
          totalContested = personHistory.length;
          totalWins = personHistory.filter((c: any) => c.is_winner).length;
          winRate = totalContested > 0 ? parseFloat(((totalWins / totalContested) * 100).toFixed(2)) : 0;
        } catch (error) {
          console.error(`Error calculating win rate for ${h.person_id}:`, error);
        }
      }

      historyRecords.push({
        year,
        winner: winnerName,
        profile_pic: personMeta.image_url || h.profile_pic,
        party: partyInfo,
        margin,
        person_id: h.person_id,
        slug: winnerName !== "Unknown" ? this.slugify(winnerName) : undefined,
        education: h.education || personMeta.education,
        profession: h.profession || personMeta.profession,
        total_assets: normalizeTotalAssets(h.total_assets || personMeta.total_assets),
        criminal_cases: normalizeCriminalCases(h.criminal_cases !== undefined ? h.criminal_cases : personMeta.criminal_cases),
        total_contested: totalContested,
        total_wins: totalWins,
        win_rate: winRate,
        age: personMeta.birth_year ? new Date().getFullYear() - parseInt(personMeta.birth_year) : (personMeta.age ? parseInt(personMeta.age) : undefined),
      });
    }

    const metadata = await this.repository.getConstituencyMetadata(constituencyId);
    const districtId = metadata?.district_id;
    const districtName = metadata?.district_name || (districtId ? districtId.replace("DISTRICT#", "").replace(/-/g, " ").replace(/\b\w/g, (l: any) => l.toUpperCase()) : undefined);

    const stats: any[] = [];
    if (metadata && metadata.statistics) {
      for (const [year, data] of Object.entries(metadata.statistics)) {
        const d = data as any;
        stats.push({
          year: parseInt(year),
          total_electors: d.total_electors || 0,
          total_votes_polled: d.total_votes_polled || 0,
          poll_percentage: d.poll_percentage || 0,
          male: d.male,
          female: d.female,
          third_gender: d.third_gender,
        });
      }
    }
    stats.sort((a, b) => b.year - a.year);

    return {
      constituency: constituencyId.replace("CONSTITUENCY#", ""),
      district_name: districtName,
      district_id: districtId,
      history: historyRecords,
      stats: stats,
    };
  }
}
